// sync-state/doc/Basic.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var React = require('react');
var jsonFormat = require('json-format');
var syncState = require('sync-state');
var Count = require('./Count');
var Range = require('./Range');
var RandomArray = require('./RandomArray');

var Basic = function (_React$Component) {
    _inherits(Basic, _React$Component);

    function Basic(props) {
        _classCallCheck(this, Basic);

        var _this = _possibleConstructorReturn(this, (Basic.__proto__ || Object.getPrototypeOf(Basic)).call(this, props));

        _this.state = {
            form: {
                name: 'abc'
            }
        };
        return _this;
    }

    _createClass(Basic, [{
        key: 'render',
        value: function render() {
            var self = this;
            var sync = syncState(this, 'form');
            return require("react").createElement(
                'div',
                null,
                require("react").createElement(
                    'pre',
                    null,
                    jsonFormat(this.state)
                ),
                require("react").createElement(
                    'h2',
                    null,
                    '\u7B80\u5355'
                ),
                require("react").createElement('input', _extends({ type: 'text' }, sync('name'))),
                require("react").createElement('input', _extends({ type: 'text' }, sync('user.title'))),
                require("react").createElement(
                    'h2',
                    null,
                    'defaultValue'
                ),
                require("react").createElement('input', _extends({
                    type: 'text'
                }, sync('some', {
                    defaultValue: 'default'
                }))),
                require("react").createElement(
                    'h2',
                    null,
                    '\u6570\u7EC4'
                ),
                require("react").createElement(RandomArray, sync('random')),
                require("react").createElement(
                    'h2',
                    null,
                    '\u522B\u540D'
                ),
                'alias: value change',
                require("react").createElement(Count, sync('age', {
                    defaultValue: 10,
                    props: 'count',
                    change: 'onMount'
                })),
                require("react").createElement(
                    'h2',
                    null,
                    '\u590D\u6742\u522B\u540D'
                ),
                require("react").createElement(Range, sync('range', {
                    defaultValue: [1, 4],
                    props: function props(value) {
                        return {
                            start: value[0],
                            end: value[1]
                        };
                    },
                    sync: function sync(start, end) {
                        return [start, end];
                    }
                }))
            );
        }
    }]);

    return Basic;
}(React.Component);
/*ONFACE-DEL*/

Basic = require("react-hot-loader").hot(module)(Basic);
module.exports = Basic;
